plenum/server/replica.py

```diff
@@ -185,6 +186,75 @@ def clear_below_view(self, view_no):
 
 
+class ConsensusDataHelper:
+    def __init__(self, consensus_data: ConsensusSharedData):
+        self.consensus_data = consensus_data
+
+    def preprepare_batch(self, pp: PrePrepare):
+        """
+        After pp had validated, it placed into _preprepared list
+        """
+        if pp in self.consensus_data.preprepared:
+            raise LogicError('New pp cannot be stored in preprepared')
+        if pp in self.consensus_data.prepared:
+            raise LogicError('New pp cannot be stored in prepared')
+        if self.consensus_data.checkpoints and pp.ppSeqNo < self.consensus_data.last_checkpoint.seqNoEnd:
+            raise LogicError('ppSeqNo cannot be lower than last checkpoint')
+        self.consensus_data.preprepared.append(pp)
+
+    def prepare_batch(self, pp: PrePrepare):
+        """
+        After prepared certificate for pp had collected,
+        it removed from _preprepared and placed into _prepared list
+        """
+        if pp not in self.consensus_data.preprepared:
+            raise LogicError('Unprepared pp must be stored in preprepared')
+        self.consensus_data.preprepared.remove(pp)
+        self.consensus_data.prepared.append(pp)
+
+    def clear_batch(self, pp: PrePrepare):
+        """
+        When 3pc batch processed, it removed from _prepared list
+        """
+        if pp in self.consensus_data.preprepared:
+            self.consensus_data.preprepared.remove(pp)
+        if pp in self.consensus_data.prepared:
+            self.consensus_data.prepared.remove(pp)
+
+    def clear_batch_till_seq_no(self, seq_no):
+        self.consensus_data.preprepared = [pp for pp in self.consensus_data.preprepared if pp.ppSeqNo < seq_no]
+        self.consensus_data.prepared = [p for p in self.consensus_data.prepared if p.ppSeqNo < seq_no]
+
+    def clear_all_batches(self):
+        """
+        Clear up all preprepared and prepared
+        """
+        self.consensus_data.prepared.clear()
+        self.consensus_data.preprepared.clear()
+
+    def add_checkpoint(self, checkpoint: Checkpoint):
+        self.consensus_data.checkpoints.add(checkpoint)
+
+    def remove_checkpoint(self, end_seq_no):
+        new_checkpoints = [c for c in self.consensus_data.checkpoints if c.seqNoEnd != end_seq_no]
+        if len(new_checkpoints) != len(self.consensus_data.checkpoints) - 1:
+            raise LogicError('One checkpoint needed to be removed')
+        self.consensus_data.checkpoints = new_checkpoints
+
+    def reset_checkpoints(self):
+        # That function most probably redundant in PBFT approach,
+        # because according to paper, checkpoints cleared only when next stabilized.
+        # Avoid using it while implement other services.
+        self.consensus_data.checkpoints.clear()
+        self.consensus_data.stable_checkpoint = None
+
+    def set_stable_checkpoint(self, end_seq_no):
+        if not self.consensus_data.checkpoints.irange_key(end_seq_no, end_seq_no):
```

Comment:
I might be wrong, but there is a chance that additional cast to list will be needed.  (irange_key might return generator, please test it at least in interactive python console)

Reply:
you are right, thanks. fixed